sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -199,45 +200,106 @@
  * <tt>"group.id"</tt>, <tt>"enable.auto.commit"</tt>, etc.
  */
 public class KafkaIO {
+  /**
+   * Creates an uninitialized {@link Read} {@link PTransform}. Before use, basic Kafka
+   * configuration should set with {@link Read#withBootstrapServers(String)} and
+   * {@link Read#withTopics(List)}. Other optional settings include key and value coders,
+   * custom timestamp and watermark functions.
+   */
+  public static Read<byte[], byte[]> readBytes() {
+    return new AutoValue_KafkaIO_Read.Builder<byte[], byte[]>()
+        .setTopics(new ArrayList<String>())
+        .setTopicPartitions(new ArrayList<TopicPartition>())
+        .setKeyCoder(ByteArrayCoder.of())
+        .setValueCoder(ByteArrayCoder.of())
+        .setConsumerFactoryFn(Read.KAFKA_9_CONSUMER_FACTORY_FN)
+        .setConsumerConfig(Read.DEFAULT_CONSUMER_PROPERTIES)
+        .setMaxNumRecords(Long.MAX_VALUE)
+        .build();
+  }
 
   /**
    * Creates an uninitialized {@link Read} {@link PTransform}. Before use, basic Kafka
    * configuration should set with {@link Read#withBootstrapServers(String)} and
    * {@link Read#withTopics(List)}. Other optional settings include key and value coders,
    * custom timestamp and watermark functions.
    */
-  public static Read<byte[], byte[]> read() {
-    return new Read<byte[], byte[]>(
-        new ArrayList<String>(),
-        new ArrayList<TopicPartition>(),
-        ByteArrayCoder.of(),
-        ByteArrayCoder.of(),
-        Read.KAFKA_9_CONSUMER_FACTORY_FN,
-        Read.DEFAULT_CONSUMER_PROPERTIES,
-        Long.MAX_VALUE,
-        null);
+  public static <K, V> Read<K, V> read() {
+    return new AutoValue_KafkaIO_Read.Builder<K, V>()
+        .setTopics(new ArrayList<String>())
+        .setTopicPartitions(new ArrayList<TopicPartition>())
+        .setConsumerFactoryFn(Read.KAFKA_9_CONSUMER_FACTORY_FN)
+        .setConsumerConfig(Read.DEFAULT_CONSUMER_PROPERTIES)
+        .setMaxNumRecords(Long.MAX_VALUE)
+        .build();
+  }
+
+  /**
+   * Creates an uninitialized {@link Write} {@link PTransform}. Before use, Kafka configuration
+   * should be set with {@link Write#withBootstrapServers(String)} and {@link Write#withTopic}
+   * along with {@link Coder}s for (optional) key and values.
+   */
+  public static Write<byte[], byte[]> writeBytes() {
```

Comment:
I looked at usages of KafkaIO on github, and nobody was using the `byte[]` version directly - people were always overriding the key and value coders. I think this version should not have existed in the first place - or more like: if we started with `KafkaIO.<KeyT, ValueT>write()`, we probably would not have added `writeBytes()`, and wouldn't have regretted that it doesn't default to bytes.

It seems acceptable to me to require specifying how to encode/decode the received data; if the user wants to deal with raw bytes directly, that's easy to do - just use `writeBytes()` (and the case when you want to have a raw key but encode the value, or vice versa, I think, is sufficiently marginal that a user can specify the ByteArrayCoder directly).

I agree having "default values" for the type parameters and, accordingly, some fields depending on them (such as coders) can be a minor convenience, but it's very hard to do in a type-safe way in general, and I think the community benefits of having an important connector using a recommended, simple-to-code, easily-reusable pattern outweigh this cost.